File: coordinator.py
import grpc
import time
import uuid
import sys
import threading
from concurrent import futures
import twopc_pb2
import twopc_pb2_grpc

import builtins
print = lambda *args, **kwargs: builtins.print(*args, flush=True, **kwargs)
SEPARATOR = "\n" + "-" * 60 + "\n"
print("[COORDINATOR] Starting process...")

PARTICIPANTS = ["participant1:50051", "participant2:50051", "participant3:50051", "participant4:50051"]
DECISION_COORDINATOR_ADDR = "localhost:60051"

class Coordinator(twopc_pb2_grpc.CoordinatorServiceServicer):
    def RequestVote(self, request, context):
        transaction_id = request.transaction_id
        print(f"[COORDINATOR] Requesting votes for Transaction {transaction_id}")
        votes = []
        for node in PARTICIPANTS:
            try:
                with grpc.insecure_channel(node) as channel:
                    stub = twopc_pb2_grpc.VotingServiceStub(channel)
                    response = stub.Vote(twopc_pb2.VoteRequest(transaction_id=transaction_id, node_id=node))
                    votes.append(response.commit)
                    print(f"[COORDINATOR] Received vote from {node}: {'Commit' if response.commit else 'Abort'}")
            except Exception as e:
                print(f"[COORDINATOR WARNING] Failed to connect to {node}: {e}")
        return twopc_pb2.VoteResponse(transaction_id=transaction_id, node_id="Coordinator", commit=all(votes))

def serve():
    print("[COORDINATOR] Initializing gRPC server...")
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    twopc_pb2_grpc.add_CoordinatorServiceServicer_to_server(Coordinator(), server)
    server.add_insecure_port("[::]:50050")
    server.start()
    print("[COORDINATOR] gRPC server is running on port 50050.")
    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        print("[COORDINATOR] Shutting down service...")
        server.stop(0)
    except Exception as e:
        print(f"[COORDINATOR WARNING] {e}", file=sys.stderr)

if __name__ == "__main__":
    print("[COORDINATOR] Executing serve() function...")
    server_thread = threading.Thread(target=serve, daemon=True)
    server_thread.start()
    while True:
        time.sleep(5)
        transaction_id = str(uuid.uuid4())
        print(SEPARATOR)
        print(f"[COORDINATOR] Sending test VoteRequest for Transaction {transaction_id}...")
        vote_reports = []
        for node in PARTICIPANTS:
            try:
                with grpc.insecure_channel(node) as channel:
                    stub = twopc_pb2_grpc.VotingServiceStub(channel)
                    response = stub.Vote(twopc_pb2.VoteRequest(transaction_id=transaction_id, node_id=node))
                    vote_reports.append(twopc_pb2.VoteReport(
                        transaction_id=transaction_id,
                        node_id=node,
                        commit=response.commit
                    ))
                    print(f"[COORDINATOR] Received vote from {node}: {'Commit' if response.commit else 'Abort'}")
            except Exception as e:
                print(f"[COORDINATOR WARNING] Could not reach {node}: {e}")
        time.sleep(1.0)
        print("[COORDINATOR] Voting phase completed.")
        decision_request = twopc_pb2.DecisionRequest(
            transaction_id=transaction_id,
            votes=vote_reports
        )
        try:
            with grpc.insecure_channel(DECISION_COORDINATOR_ADDR) as channel:
                stub = twopc_pb2_grpc.DecisionServiceStub(channel)
                print("Phase Voting of Node coordinator sends RPC StartDecisionPhase to Phase Decision of Node coordinator")
                ack = stub.StartDecisionPhase(decision_request)
                time.sleep(0.5)
                global_commit = all(v.commit for v in vote_reports)
                decision_text = "COMMITTED" if global_commit else "ABORTED"
                print(f"[COORDINATOR] Global decision for {transaction_id}: {decision_text}")
                print(f"[COORDINATOR] Got ack from Go coordinator: {ack.message}")
        except Exception as e:
            print(f"[COORDINATOR WARNING] Failed to reach Go decision coordinator: {e}")
        time.sleep(1.5)

File: coordinator.go
package main
import (
	"context"
	"log"
	"net"
	"os"
	"time"
	pb "twopcapp/proto"
	"google.golang.org/grpc"
)
type ParticipantServer struct {
	pb.UnimplementedDecisionServiceServer
	pb.UnimplementedGlobalDecisionServiceServer
	nodeID   string
	lastVote bool
}
func (s *ParticipantServer) ReportVote(ctx context.Context, req *pb.VoteReport) (*pb.Ack, error) {
	log.Printf("Phase Voting of Node %s sends RPC ReportVote to Phase Decision of Node %s", req.NodeId, s.nodeID)
	log.Printf("[PHASE VOTE] Node %s received ReportVote from %s: Commit=%v", s.nodeID, req.NodeId, req.Commit)
	s.lastVote = req.Commit
	return &pb.Ack{Message: "Vote recorded"}, nil
}
func (s *ParticipantServer) GlobalDecision(ctx context.Context, req *pb.DecisionMessage) (*pb.Ack, error) {
	outcome := "ABORTED"
	if req.GlobalCommit {
		outcome = "COMMITTED"
	}
	log.Printf("Phase Decision of Node coordinator sends RPC GlobalDecision to Phase Decision of Node %s", s.nodeID)
	log.Printf("[PHASE DECISION] Node %s received GlobalDecision for Tx %s => %s", s.nodeID, req.TransactionId, outcome)
	return &pb.Ack{Message: "Decision applied"}, nil
}
func (s *ParticipantServer) StartDecisionPhase(ctx context.Context, req *pb.DecisionRequest) (*pb.Ack, error) {
	log.Printf("Phase Voting of Node coordinator sends RPC StartDecisionPhase to Phase Decision of Node %s", s.nodeID)
	globalCommit := true
	for _, vote := range req.Votes {
		if !vote.Commit {
			globalCommit = false
			break
		}
	}
	log.Printf("[COORDINATOR] Global decision for %s: %v", req.TransactionId, globalCommit)
	for _, vote := range req.Votes {
		host, _, err := net.SplitHostPort(vote.NodeId)
		if err != nil {
			log.Printf("[COORDINATOR ERROR] Invalid node ID format: %s", vote.NodeId)
			continue
		}
		address := host + ":60051"
		conn, err := grpc.Dial(address, grpc.WithInsecure())
		if err != nil {
			log.Printf("[COORDINATOR ERROR] Could not reach %s: %v", address, err)
			continue
		}
		client := pb.NewGlobalDecisionServiceClient(conn)
		_, err = client.GlobalDecision(context.Background(), &pb.DecisionMessage{
			TransactionId: req.TransactionId,
			GlobalCommit:  globalCommit,
		})
		if err != nil {
			log.Printf("[COORDINATOR ERROR] Sending GlobalDecision to %s failed: %v", address, err)
		} else {
			log.Printf("[COORDINATOR] Decision sent to %s: %s", address, map[bool]string{true: "COMMIT", false: "ABORT"}[globalCommit])
		}
		conn.Close()
	}
	return &pb.Ack{Message: "Global decision made"}, nil
}
func main() {
	log.SetFlags(0)
	log.SetOutput(os.Stdout)
	nodeID := os.Getenv("NODE_ID")
	if nodeID == "" {
		log.Fatal("NODE_ID environment variable not set")
	}
	port := ":60051"
	lis, err := net.Listen("tcp", port)
	if err != nil {
		log.Fatalf("Failed to listen: %v", err)
	}
	grpcServer := grpc.NewServer()
	server := &ParticipantServer{nodeID: nodeID}
	pb.RegisterDecisionServiceServer(grpcServer, server)
	pb.RegisterGlobalDecisionServiceServer(grpcServer, server)
	log.Printf("[%s] Phase Decision of Node %s listening on %s...", time.Now().Format(time.RFC3339Nano), nodeID, port)
	if err := grpcServer.Serve(lis); err != nil {
		log.Fatalf("Failed to serve: %v", err)
	}
}
